lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -84,17 +179,114 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
     if (Replacement.findIf([](Type ty) -> bool {
           return getBoundGenericDepth(ty) >= BoundGenericDepthThreshold;
         })) {
-      return;
+      DEBUG(llvm::dbgs()
+            << "    Cannot specialize because the generic type is too deep.\n");
+      return false;
     }
   }
 
-  SILModule &M = OrigF->getModule();
-  SubstitutedType = SILType::substFuncType(M, InterfaceSubs,
-                                           OrigF->getLoweredFunctionType(),
-                                           /*dropGenerics = */ true);
+  // Check if we have substitutions which replace generic type parameters with
+  // concrete types or unbound generic types.
+  bool HasConcreteGenericParams = false;
+  HasUnboundGenericParams = false;
+  for (auto &entry : InterfaceSubs.getMap()) {
+    // Check only the substitutions for the generic parameters.
+    // Ignore any dependent types, etc.
+    if (!isa<GenericTypeParamType>(entry.first))
+      continue;
+    auto Replacement = entry.second->getCanonicalType();
+    if (Replacement->hasArchetype()) {
+      HasUnboundGenericParams = true;
+      continue;
+    }
+    HasConcreteGenericParams = true;
+  }
+
+  if (!EnablePartialSpecialization && HasUnboundGenericParams) {
+    DEBUG(llvm::dbgs() <<
+          "    Partial specialization is not enabled.\n");
+    return false;
+  }
 
+#if 0
+  if (!HasConcreteGenericParams) {
+    // All substititions are unbound.
+    DEBUG(llvm::dbgs() <<
+          "    Cannot specialize with all unbound interface substitutions.\n");
+    DEBUG(for (auto Sub : ParamSubs) {
+            Sub.dump();
+          });
+    return false;
+  }
+#endif
+
+  // Do not partially specialize any print.*unlock methods, because
+  // they are very big and are not important for performance.
+  // TODO: Could it be that they are important for string interpolation?
+  // TODO: Introduce a proper way to tell the compiler that certain
+  // functions should not be (auto)-(partially)-specialized.
+  if (Callee->getName().find("_unlock", 0) != StringRef::npos)
+    return false;
+
+  // We need a generic environment for the partial specialization.
+  if (HasUnboundGenericParams && !Callee->getGenericEnvironment())
```

Comment:
If the generic callee doesn't have a generic environment, it has no body -- so you won't be specializing it. This check is unnecessary